### ISSUE TITLE
Zero volume ellipsoid enclosing non-zero volume cube.

### ISSUE THREAD

**Issue by n3vu0r** [CGAL/cgal#6626]
## Issue Details

We provide 8 corner points of a 3D cube and verify that all points have the same non-zero distance from the cube's center by asserting the distances have a non-zero average with vanishing standard deviation. Then we compute the minimum volume ellipsoid enclosing the cube and assert that it is full dimensional or equally that it has a non-zero volume. This assertions fails with:

```text
int main(): Assertion `ellipsoid.is_full_dimensional()' failed.
```

Translating the cube randomly will result in failing and succeeding assertions. It seems `is_full_dimensional()` is numerical unstable.

## Source Code

```c++
#include <vector>
#include <CGAL/Cartesian_d.h>
#include <CGAL/MP_Float.h>
#include <CGAL/point_generators_d.h>
#include <CGAL/Approximate_min_ellipsoid_d.h>
#include <CGAL/Approximate_min_ellipsoid_d_traits_d.h>
#include "stdio.h"

typedef CGAL::Cartesian_d<double> Kernel;
typedef CGAL::MP_Float ET;
typedef CGAL::Approximate_min_ellipsoid_d_traits_d<Kernel, ET> Traits;
typedef Traits::Point Point;
typedef CGAL::Approximate_min_ellipsoid_d<Traits> Ellipsoid;

// Cube of non-zero volume (8 corner points).
std::vector<Point> cube() {
	// 8 corner points.
	double l[3] = { 390,  971, 8957 }; // lower (0)
	double u[3] = { 490, 1071, 9057 }; // upper (1)
	double p[8][3] = {
		{ l[0], l[1], l[2] }, // (0, 0, 0)
		{ u[0], l[1], l[2] }, // (1, 0, 0)
		{ u[0], u[1], l[2] }, // (1, 1, 0)
		{ l[0], u[1], l[2] }, // (0, 1, 0)
		{ l[0], l[1], u[2] }, // (0, 0, 1)
		{ u[0], l[1], u[2] }, // (1, 0, 1)
		{ u[0], u[1], u[2] }, // (1, 1, 1)
		{ l[0], u[1], u[2] }, // (0, 1, 1)
	};
	std::vector<Point> points;
	for (size_t i = 0; i < 8; ++i)
		points.push_back(Point(3, std::begin(p[i]), std::end(p[i])));
	// Center of cube.
	double center[3] = { 0, 0, 0 };
	for (size_t i = 0; i < points.size(); ++i) {
		center[0] += points[i][0];
		center[1] += points[i][1];
		center[2] += points[i][2];
	}
	center[0] /= points.size();
	center[1] /= points.size();
	center[2] /= points.size();
	// Point-center distances, their average and standard deviation.
	double distances[8];
	double distances_avg = 0;
	for (size_t i = 0; i < points.size(); ++i) {
		distances[i] = sqrt(
			pow(points[i][0] - center[0], 2) +
			pow(points[i][1] - center[1], 2) +
			pow(points[i][2] - center[2], 2)
		);
		distances_avg += distances[i];
	}
	distances_avg /= points.size();
	// Assert that cube has non-zero volume.
	assert(distances_avg == 86.602540378443877);
	double distances_err = 0;
	for (size_t i = 0; i < points.size(); ++i) {
		distances_err = pow(distances[i] - distances_avg, 2);
	}
	distances_err /= points.size();
	distances_err = sqrt(distances_err);
	// Assert that points describe cube.
	assert(distances_err == 5.0242958677880805e-15);
	return points;
}

// Assertion fails.
int main() {
	// Cube of non-zero volume (8 corner points).
	auto points = cube();
	// Ellipsoid enclosing cube.
	Ellipsoid ellipsoid(0.01, points.begin(), points.end(), Traits());
	// Assert that ellipsoid enclosing non-zero volume cube has non-zero volume.
	assert(ellipsoid.is_full_dimensional());
}
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): `Arch Linux 64 bit`
* Compiler: `g++ (GCC) 12.1.0`
* Release or debug mode: Default of `CMAKE_BUILD_TYPE`
* Specific flags used (if any): `-O2 -Werror -Wall -Wextra -pedantic`
* CGAL version: `5.4`
* Boost version: `1.78.0`
* Other libraries versions if used (Eigen, TBB, etc.): `eigen 3.4.0`

**Comment by n3vu0r** [comment]
To clarify, all edge lengths of the cube are `100.0`, so the volume is large with `1e6`. Thus, it's not the rounding problem described in the documentation regarding `is_fulll_dimensional()`. It seems that as soon as two semi-axis lengths become close, the algorithm fails. It means, the semi-axes directions are no longer unique but they still span a large full dimensional space.

**Comment by afabri** [comment]
Hello, @n3vu0r we have a PR. It will go through the CGAL testsuite and be later merged into the master branch, but as the fix is small you might apply it locally so that you can go on with what you work on.  Thank you for the bug report, as only feedback enables us to improve.

**Comment by n3vu0r** [comment]
Removing line 300 in `/usr/include/CGAL/Approximate_min_ellipsoid_d.h` works for me. Is it safe to do so?

**Comment by afabri** [comment]
Bernd told me it is safe.   
To our excuse I must say that this is not a heavily used package.   If you had test cases that you consider as hard we would be glad to get them, so keep on posting when you hit a problem.   
Do you by any chance have code to generate a 3D triangle mesh for 3D ellipsoids?     
In your code you used the dD kernel with 3D points, Are you aware that there is also a 3D traits class?

**Comment by n3vu0r** [comment]
I will post when I hit a problem or have further tests. So far it works for a large number of small point sets. But it doesn't include degenerate cases (my points must always span larger volumes).

Sorry, I don't have code for 3D triangle meshes myself, I'm calling CGAL from Rust and currently use [`kiss3d`](https://github.com/sebcrozet/kiss3d) to draw. I started with the dD example and switched now to the 3D kernel, thanks.

> ...enclosing ellipsoid whose volume is by at most a factor of 1+ϵ larger than the volume of the smallest enclosing ellipsoid...

When 1+ϵ is only an upper limit factor for the volume, it should be fine to enlarge negative epsilon to zero. But shouldn't the documentation of the post condition for `achieved_epsilon()` be updated from `ϵ > 0` to `ϵ >= 0`? This might be overall inconsistent in the documentation, not sure.

I've tested following changes and it works as well for my point sets:

Clamping epsilon in `exact_epsilon()`:
```c++
    if (CGAL::is_negative(eps_e)) {
      CGAL_APPEL_LOG("appel", "Negative Exact epsilon -> clamp to zero!" << "\n");
      eps_exact = 0.0;
    }
```
Enabling again the post-condition check of `achieved_epsilon()`:

```c++
      CGAL_APPEL_ASSERT(eps >= 0.0);
```
